parser/analyzer.go

```diff
@@ -0,0 +1,217 @@
+package parser
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/pkg/errors"
+
+	"github.com/ulule/loukoum/lexer"
+	"github.com/ulule/loukoum/token"
+)
+
+// ErrAnalyzer is returned when analyze was incomplete.
+var ErrAnalyzer = fmt.Errorf("analyze was incomplete")
+
+// AnalyzerOption defines what should be analyzed while scanning statements.
+type AnalyzerOption struct {
+	// Operation inspects what operation (or command) is performed.
+	// Example: INSERT, UPDATE, SELECT, DELETE, etc...
+	Operation bool
+	// Table inspects what table will handle the given statement.
+	Table bool
+}
+
+// Continue determines if we should keep scanning statements, or if we have everything we need.
+func (o AnalyzerOption) Continue(result *AnalyzerResult) bool {
+	if o.Operation && result.Operation == "" {
+		return true
+	}
+	if o.Table && result.Table == "" {
+		return true
+	}
+	return false
+}
+
+// AnalyzerResult is the result produces by an analyzer with the given options.
+type AnalyzerResult struct {
+	// Operation defines the operation (or command).
+	// Example: INSERT, UPDATE, SELECT, DELETE, etc...
+	Operation string
+	// Table defines the table of given statement.
+	Table string
+}
+
+// Analyze will analyzes given query with options.
+func Analyze(query string, option AnalyzerOption) (*AnalyzerResult, error) { // nolint: gocyclo
+	lexer := lexer.New(strings.NewReader(query))
+	it := lexer.Iterator()
+
+	mode := token.Illegal
+	result := &AnalyzerResult{}
+
+	if !option.Continue(result) {
+		return result, nil
+	}
+
+	for it.HasNext() {
+		e := it.Next()
+		switch e.Type {
+		case token.Select:
+			if mode == token.Illegal {
+				mode = e.Type
+			}
+			if !option.Operation {
+				continue
+			}
+
+			result.Operation = e.Type.String()
+			if !option.Continue(result) {
```

Comment:
maybe reverse this thing using `result.Filled()` to keep an option as an option without operation

Reply:
The option is mandatory, but I could change that to `result.Finished(option)` if you prefer.

Reply:
Maybe option is a bad wording and we should change that to `query` or anything else.